### PR TITLE
Descendant users

### LINKED ISSUE #188: Group composition: 4. descendant users
## Motivations
List composition of a group like in https://france-ioi.github.io/algorea-designs/01.Groups_02.Composition

## Subtasks
* Create the children/descendant filter
* Do not display any second filter
* Use service: https://france-ioi.github.io/algorea-devdoc/api/#operation/groupUserDescendantView
* Display column:
  - name (firstname+lastname)
  - login
  - parent group(s): comma separated list of parent group's name.
* do not allow any sort

### PR DESCRIPTION
**Pull request by GabrielVidal1**
Fixes #188

### CODE REVIEW

File: src/app/modules/group/components/member-list/member-list.component.html [hunk 1]
```diff
@@ -54,6 +54,12 @@
           <ng-container *ngIf="col.field === 'user_count'">
             {{ rowData.userCount }}
           </ng-container>
+          <ng-container *ngIf="col.field === 'login'">
+            {{ rowData.login }}
+          </ng-container>
+          <ng-container *ngIf="col.field === 'parentGroups'">
```
Comment (smadbe) [inline src/app/modules/group/components/member-list/member-list.component.html:60]: it becomes quite repetitive, you could do a switch (or if-elseif-else) on col.field, treat 'user.login', 'member_since' and 'user_count' separately, then on default do `{{ rowData[col.field] }}`

File: src/app/modules/group/components/group-composition-filter/group-composition-filter.component.ts [hunk 2]
```diff
@@ -22,29 +23,53 @@ export class GroupCompositionFilterComponent implements OnInit{
...
-      icon: 'fa fa-calendar',
-      label: 'sessions',
-      type: TypeFilter.Sessions
+      icon: 'fa fa-user',
+      label: 'users',
+      value: TypeFilter.Users
     },
+  ]
+
+  directChildrenTypeFilters = [
```
[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/modules/group/components/group-composition-filter/group-composition-filter.component.ts:52]: Ah I got why you did it... to keep the selection of "users" when going from one to the other... But that's not very clean as reordering the filter (a UI thing) will be break this mechanism. You could instead, when switching, search the value in the other table and fallback on default for instance.

File: src/app/modules/group/components/member-list/member-list.component.ts [hunk 3]
```diff
@@ -42,7 +48,7 @@ export class MemberListComponent implements OnChanges, OnDestroy {
 
   state: 'error' | 'ready' | 'fetching' = 'fetching';
 
-  defaultPolicy: Filter = { type: TypeFilter.Groups };
+  defaultPolicy: Filter = { type: TypeFilter.Groups, directChildren: false };
```
Comment (smadbe) [inline src/app/modules/group/components/member-list/member-list.component.ts:51]: There are still several usage of "policy" in this file...

File: src/app/modules/group/components/group-composition-filter/group-composition-filter.component.html [hunk 4]
```diff
@@ -1,8 +1,13 @@
 <div class="selection-tool">
+  <alg-selection style="margin: 10px;"
+    [items]="childrenFilters"
+    [default]="+value.directChildren"
```
[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/modules/group/components/group-composition-filter/group-composition-filter.component.html:4]: And now I saw the rest of the code, I think you also do the assumption that `defaultPolicy` of the parent component is the value. It is also an assumption which is not good to take. The default should be at one single place and passed/computed.

File: src/app/shared/http-services/get-group-user-descendants.service.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  name: string,
+  parents: {
+    id: string,
+    name: string,
+  }[],
+  user: {
+    login: string,
+    firstName: string|null,
+    lastName: string|null,
+    grade: number,
```
Comment (smadbe) [inline src/app/shared/http-services/get-group-user-descendants.service.ts:33]: grade is nullable!

**Review by smadbe — CHANGES_REQUESTED** [review]

File: src/app/shared/http-services/get-group-user-descendants.service.ts (hunk 5 above)
Comment (GabrielVidal1) [reply src/app/shared/http-services/get-group-user-descendants.service.ts]: fixed

File: src/app/modules/group/components/member-list/member-list.component.ts (hunk 3 above)
Comment (GabrielVidal1) [reply src/app/modules/group/components/member-list/member-list.component.ts]: fixed

File: src/app/modules/group/components/group-composition-filter/group-composition-filter.component.ts [hunk 6]
```diff
@@ -22,29 +23,53 @@ export class GroupCompositionFilterComponent implements OnInit{
 
   @Output() change = new EventEmitter<Filter>();
 
-  value: Filter = { type: TypeFilter.Users };
+  value: Filter = { type: TypeFilter.Users, directChildren: true };
 
-  typePolicies = [
+  childrenFilters: { label:string, value: boolean }[] = [
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/group/components/group-composition-filter/group-composition-filter.component.ts]: fixed

File: src/app/modules/group/components/group-composition-filter/group-composition-filter.component.ts [hunk 7]
```diff
@@ -22,29 +23,53 @@ export class GroupCompositionFilterComponent implements OnInit{
...
     },
     {
-      icon: 'fa fa-calendar',
-      label: 'sessions',
-      type: TypeFilter.Sessions
+      icon: 'fa fa-user',
+      label: 'users',
+      value: TypeFilter.Users
     },
+  ]
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/group/components/group-composition-filter/group-composition-filter.component.ts]: fixed

File: src/app/modules/group/components/member-list/member-list.component.html (hunk 1 above)
Comment (GabrielVidal1) [reply src/app/modules/group/components/member-list/member-list.component.html]: fixed

File: src/app/modules/group/components/group-composition-filter/group-composition-filter.component.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/group/components/group-composition-filter/group-composition-filter.component.ts]: fixed

File: src/app/modules/group/components/group-composition-filter/group-composition-filter.component.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/group/components/group-composition-filter/group-composition-filter.component.html]: fixed

File: src/app/modules/shared-components/components/selection/selection.component.ts [hunk 8]
```diff
@@ -9,13 +9,16 @@ export class SelectionComponent implements OnInit {
...
   @Input() mode: 'light' | 'dark' | 'basic' = 'light';
 
   @Output() change = new EventEmitter<number>();
 
   constructor() {}
 
-  ngOnInit(): void {}
+  ngOnInit(): void {
+    if (this.default) this.selected = this.default;
+  }
```
Comment (smadbe) [inline src/app/modules/shared-components/components/selection/selection.component.ts:22]: but do you do not need this I assume

**Review by smadbe — CHANGES_REQUESTED** [review]
Looks like the default selection is quite wrong... 
Screenshot 2020-11-25 at 18 18 15

File: src/app/modules/shared-components/components/selection/selection.component.ts (hunk 8 above)
Comment (GabrielVidal1) [reply src/app/modules/shared-components/components/selection/selection.component.ts]: fixed

**Comment by GabrielVidal1** [comment]
I just need to rebase on master to fix the e2e tests

**Review by smadbe — CHANGES_REQUESTED** [review]
When switching children<->descendants, the type is set to Users. While when first arriving on the page it is "Teams"... bit strange/smelly.
Could you:
* When the current type is teams or users, keep the same type when switching children<->descendants 
* When landing on the page: select direct users
* When switching children<->descendants with subgroups or sessions selected, choose "Users" (good now)

**Comment by GabrielVidal1** [comment]
Ok, I fixed it.

**Review by smadbe — APPROVED** [review]
Looks good 👍

**Pull request merged**